### PR TITLE
Avoding crashes when broker reconnects

### LINKED ISSUE #18: reconnection problem when broker is restarted
I have written an app that uses Tortoise to connect to a mosquitto broker(v2.0.14). Everything works fine until I stop the broker. At first I used publish directly from the place where the mqtt messages are created. Each time the broker is stopped, and then restarted, the messages in the inflight are sent. But then the client (Tortoise) closes the connection again and never recovers.

I get these errors in the console. First (only once)

15:56:31.059 [error] GenStateMachine {Tortoise311.Registry, {Tortoise311.Connection.Inflight, "HeinSim"}} terminating
** (exit) {:protocol_violation, :unknown_identifier}
    (stdlib 3.16.1) gen_statem.erl:1388: :gen_statem.loop_state_callback_result/11
    (stdlib 3.16.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
State: {{:connected, {Tortoise311.Transport.Tcp, #Port<0.14>}}, %Tortoise311.Connection.Inflight{client_id: "HeinSim", order: [], pending: %{}}}
Callback mode: :handle_event_function
Postponed events: []
Last event: {:cast, {:update, {:received, %Tortoise311.Package.Puback{__META__: %Tortoise311.Package.Meta{flags: 0, opcode: 4}, identifier: 27484}}}}
Queued events: []
State: {{:connected, {Tortoise311.Transport.Tcp, #Port<0.14>}}, %Tortoise311.Connection.Inflight{client_id: "HeinSim", order: [], pending: %{}}}

and then ( for each message in inflight)

15:56:31.416 [error] GenStateMachine {Tortoise311.Registry, {Tortoise311.Connection.Inflight, "HeinSim"}} terminating
** (CaseClauseError) no case clause matching: {:error, :closed}
    (tortoise311 0.11.1) lib/tortoise311/connection/inflight.ex:292: Tortoise311.Connection.Inflight.handle_event/4
    (stdlib 3.16.1) gen_statem.erl:1196: :gen_statem.loop_state_callback/11
    (stdlib 3.16.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
State: {{:connected, {Tortoise311.Transport.Tcp, #Port<0.14>}}, %Tortoise311.Connection.Inflight{client_id: "HeinSim", order: [17311], pending: %{17311 => %Tortoise311.Connection.Inflight.Track{caller: {#PID<0.316.0>, #Reference<0.3953562195.1766850567.237007>}, identifier: 17311, pending: [[dispatch: %Tortoise311.Package.Publish{__META__: %Tortoise311.Package.Meta{flags: 0, opcode: 3}, dup: false, identifier: 17311, payload: "....", qos: 1, retain: false, topic: "BH/1/670820"}, received: %Tortoise311.Package.Puback{__META__: %Tortoise311.Package.Meta{flags: 0, opcode: 4}, identifier: 17311}], [{:respond, {#PID<0.316.0>, #Reference<0.3953562195.1766850567.237007>}}, :cleanup]], polarity: :negative, status: [], type: Tortoise311.Package.Publish}}}}
Callback mode: :handle_event_function
Postponed events: []

Since the connection is setup again after the broker restarts, some messages go through, but then the connection is closed almost immediately:

1639148191: New client connected from 127.0.0.1:56500 as HeinSim (p2, c0, k60, u'u1').
1639148191: Client HeinSim closed its connection.

And it never recovers.

I then changed the implementation to add a queue and process the messages one by one via publish_sync, but with the same result. After the broker comes up again, the connections is restored but almost immediately closed, and never recovers.

The Tortoise connection is supervised by my app supervisor.

I am sure that I do something wrong, but I have no idea what. Is it because of the client-id? Hopefully someone can help me.

### PR DESCRIPTION
**Pull request by jfcloutier**
Fixes #18

### CODE REVIEW

**Review by jjcarstens — APPROVED** [review]

File: lib/tortoise311/connection/inflight.ex [hunk 1]
```diff
@@ -244,7 +246,8 @@ defmodule Tortoise311.Connection.Inflight do
       {:keep_state, data, next_actions}
     else
       :error ->
-        {:stop, {:protocol_violation, :unknown_identifier}, data}
+        Logger.warn("[Tortoise311] Unknown identifier in #{inspect(update)}. Ignoring it.")
+        :keep_state_and_data
```
Comment (fhunleth) [inline lib/tortoise311/connection/inflight.ex:250]: If the client is out of sync with the server, shouldn't Tortoise311 close the socket and reconnect? It seems like this should be unrecoverable.

**Comment by fhunleth** [comment]
It seems like this PR does improve the situation, but I feel like there's a more general fix to recover from any crash to `Inflight`. Skimming the code, it looks like other lines could crash as well. Given the bug report, would it be correct in assuming that those would not be handled either?

File: lib/tortoise311/connection/inflight.ex (hunk 1 above)
Comment (jjcarstens) [reply lib/tortoise311/connection/inflight.ex:250]: I just did a quick breeze of MQTT spec to compare data from the bug report. It looks like the initial data that cause the crash was opcode 4 with identifier 27484. I think that's a PUBACK which seems like something we should actually be handling

I guess I was viewing this PR as getting progress growing but may need to be adapted

**Comment by jfcloutier** [comment]
> It seems like this PR does improve the situation, but I feel like there's a more general fix to recover from any crash to `Inflight`. Skimming the code, it looks like other lines could crash as well. Given the bug report, would it be correct in assuming that those would not be handled either?

Inflight is meant to stop if an update to a track (a back and forth wth the MQTT server) can not be resolved, i.e. a protocol error occurred. 

The one error that is now ignored is when the MQTT server sends a package for a track that is no longer "in flight" because the Inflight was somehow just reset. Crashing the Inflight had the MQTT server retrying, causing another crash, etc.

I can't think of a better solution than to ignore a package directed at a "track" that has vanished, but I am open to one.

File: lib/tortoise311/connection/inflight.ex (hunk 1 above)
Comment (jfcloutier) [reply lib/tortoise311/connection/inflight.ex:250]: > I just did a quick breeze of MQTT spec to compare data from the bug report. It looks like the initial data that cause the crash was opcode 4 with identifier 27484. I think that's a PUBACK which seems like something we should actually be handling

PUBACKs are handled. The issue here a PUBACK sent by the MQTT server was aimed at a conversation (a Track) that had vanished. In this case, the Inflight had been reset (or restarted) and all Tracks had vanished.

Comment (jfcloutier) [reply lib/tortoise311/connection/inflight.ex:250]: > If the client is out of sync with the server, shouldn't Tortoise311 close the socket and reconnect? It seems like this should be unrecoverable.

I am not entirely sure what is best. Would the MQTT server, upon reconnection, send the same package aimed at a non-existent track, thus causing a repeating reconnect? Ignoring the mis-directed package seems the lesser evil at the moment.

File: lib/tortoise311/connection/inflight.ex [hunk 2]
```diff
@@ -244,7 +246,8 @@ defmodule Tortoise311.Connection.Inflight do
       {:keep_state, data, next_actions}
     else
       :error ->
-        {:stop, {:protocol_violation, :unknown_identifier}, data}
+        Logger.warn("[Tortoise311] Unknown identifier in #{inspect(update)}. Ignoring it.")
```
Comment (fhunleth) [inline lib/tortoise311/connection/inflight.ex:249]:
```suggestion
        Logger.warn("[Tortoise311] Ignoring response with unknown identifier: #{inspect(update)}")
```

**Review by fhunleth — APPROVED** [review]

**Pull request merged**